File: server/analysis/r/mediation.R
#!/usr/bin/env Rscript
# Mediation analysis — paths a, b, c', indirect effect, bootstrap CI
suppressPackageStartupMessages({
  library(jsonlite)
  library(lavaan)
})

input <- readLines(file("stdin"), warn = FALSE)
parsed <- fromJSON(paste(input, collapse = "\n"))
data <- as.data.frame(parsed$data)
config <- parsed$config

x <- config$x
m <- config$m
y <- config$y

model <- paste0(
  m, " ~ a*", x, "\n",
  y, " ~ b*", m, " + cp*", x, "\n",
  "indirect := a*b"
)

fit <- sem(model, data = data, se = "bootstrap", bootstrap = 500)
pe <- parameterEstimates(fit)

get_par <- function(lbl) {
  r <- pe[pe$label == lbl, ]
  if (nrow(r) == 0) return NULL
  list(est = r$est, se = r$se, pvalue = r$pvalue, ci_lower = r$ci.lower, ci_upper = r$ci.upper)
}

indirect <- pe[pe$label == "indirect", ]
if (nrow(indirect) == 0) {
  indirect <- tryCatch({
    indirect_est <- pe[pe$label == "a", "est"] * pe[pe$label == "b", "est"]
    list(est = indirect_est, se = NA, pvalue = NA, ci_lower = NA, ci_upper = NA)
  }, error = function(e) list(est = NA, se = NA, pvalue = NA, ci_lower = NA, ci_upper = NA))
} else {
  indirect <- as.list(indirect[1, c("est", "se", "pvalue", "ci.lower", "ci.upper")])
  names(indirect) <- c("est", "se", "pvalue", "ci_lower", "ci_upper")
}

total_c <- get_par("cp")
if (is.null(total_c)) total_c <- list(est = pe[pe$rhs == x & pe$lhs == y, "est"][1], se = NA, pvalue = NA)

out <- list(
  path_a = get_par("a"),
  path_b = get_par("b"),
  path_c_prime = get_par("cp"),
  indirect_effect = indirect,
  total_effect = total_c
)
cat(toJSON(out, auto_unbox = TRUE, digits = 8))

File: server/analysis/python/mediation.py
#!/usr/bin/env python3
"""Mediation analysis — paths a, b, c', indirect effect, bootstrap CI."""
import json
import sys
import pandas as pd
import numpy as np
import statsmodels.api as sm

input_json = sys.stdin.read()
parsed = json.loads(input_json)
data = pd.DataFrame(parsed["data"])
config = parsed["config"]

x_col = config["x"]
m_col = config["m"]
y_col = config["y"]

n_boot = 500


def run_paths(df):
    X_m = sm.add_constant(df[x_col])
    model_m = sm.OLS(df[m_col], X_m).fit()
    a = model_m.params[x_col]

    X_y = sm.add_constant(df[[m_col, x_col]])
    model_y = sm.OLS(df[y_col], X_y).fit()
    b = model_y.params[m_col]
    cp = model_y.params[x_col]
    return a, b, cp


a, b, cp = run_paths(data)
indirect_obs = a * b

boot_indirect = []
np.random.seed(42)
n = len(data)
for _ in range(n_boot):
    idx = np.random.choice(n, n, replace=True)
    sample = data.iloc[idx]
    try:
        ai, bi, _ = run_paths(sample)
        boot_indirect.append(ai * bi)
    except Exception:
        pass

boot_indirect = np.array(boot_indirect)
ci_lower = float(np.percentile(boot_indirect, 2.5)) if len(boot_indirect) else None
ci_upper = float(np.percentile(boot_indirect, 97.5)) if len(boot_indirect) else None
se_boot = float(np.std(boot_indirect)) if len(boot_indirect) else None
pval = float(np.mean(boot_indirect <= 0) * 2) if len(boot_indirect) else None

out = {
    "path_a": {"est": float(a)},
    "path_b": {"est": float(b)},
    "path_c_prime": {"est": float(cp)},
    "indirect_effect": {
        "est": float(indirect_obs),
        "se": se_boot,
        "pvalue": pval,
        "ci_lower": ci_lower,
        "ci_upper": ci_upper,
    },
    "total_effect": {"est": float(cp + indirect_obs)},
}
print(json.dumps(out, allow_nan=False))
